Update calendar code

sairamya93 opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Adds locale values for When do the clock change file, moving them from the JSON files, and remove the hack file that handled returning either a translated key (for keys in the locale files), or the original file (for hardcoded text)

We also remove the task for publishing the calendars from frontend and add the relevant data into special route publisher (here:  https://github.com/alphagov/special-route-publisher/pull/351). In order to do that, we make the description of when-do-the-clocks-change more generic (removing the list of years), and remove the body (since the body is not present for the bank holiday pages, and appears not to be used in search).

- Also sets locale correctly in ICS files, and serves Welsh content in ICS if accessed from welsh page.
- Also removes an unused bank holiday generator task

## Why

[Trello card?] https://trello.com/c/dx543EiX/322-tidy-when-do-the-clocks-change-files, [Jira issue PNP-8592](https://gov-uk.atlassian.net/browse/PNP-8592)

## How

We should handle the when do the clocks change page the same way as the bank holidays one, which would allow us to translate it accordingly in the config file and lets us remove the hack file

KludgeKML committed:
Remove unused generator task

- This is out of date with how the bank holiday files work now (and how they've worked in the past), now we're given lists of bank  holidays by the content people.

KludgeKML committed:
Update ICS Renderer to show correct locale

- ICS files now return Welsh content when accessed through the Welsh pages,
  so we should indicate the correct locale in the file header.

sairamya93 committed:
Remove I18n Hack

    - This hack captured missing translations and replaced them
      with the original key, and was used to support code from
      calendars which needed to handle either a translation key
      or hard-coded text, but this lead to surprising behaviour

sairamya93 committed:
Move data to translation keys

    - create translation keys for when-do-the-clocks-change content
    - update division slugs in both .json files to be translation keys
    - remove body (body was not used, and bank holidays body
      was empty, so no point it being there for either calendar type)
    - simplify description to remove years range, so that it doesn't
      have to be updated (important because it will be in two places,
      here and in the special-route-publisher data for this route).

sairamya93 committed:
Update tests

    - Fixtures moved to use translation keys rather than hard-coded values
    - Tests now use clocks go forward rather than clocks advance to match content
    - Tests with arbitrary division / event names now use valid names

leenagupte commented on app/controllers/calendar_controller.rb:
```diff
@@ -13,9 +13,9 @@ class InvalidCalendarScope < StandardError; end
     respond_to do |format|
       format.html do
-        @faq_presenter = FaqPresenter.new(scope, calendar, content_item_hash, view_context)
+        @faq_presenter = FaqPresenter.new(Calendar.file_name(params[:scope]), calendar, content_item_hash, view_context)
```
Why `FaqPresenter` being instantiated with `Calendar.file_name(params[:scope])` now rather than `calendar`? Are they different?

KludgeKML replied:
Fixed to return calendar.scope

leenagupte commented on app/models/calendar.rb:
```diff
@@ -14,21 +14,34 @@ def self.find(slug)
 
-  attr_reader :slug, :title, :description
+  def self.file_name(slug)
+    slug == "gwyliau-banc" ? "bank-holidays" : slug
```
It feels like this should return the file name rather than the slug.

KludgeKML replied:
Have reengineered to return "scope", which is how the bulk of the code refers to "bank-holidays" vs "when-the-clocks-change" regardless of translation.

leenagupte commented on app/models/calendar.rb:
```diff
@@ -14,21 +14,34 @@ def self.find(slug)
+    slug == "gwyliau-banc" ? "bank-holidays" : slug
+  end
+
+  def years_covered
```
I think this method is missing a test.

KludgeKML replied:
Method no longer exists, so marking this as resolved.

leenagupte reviewed (commented):
@KludgeKML, @sairamya93 This looks like a worthwhile change, and thanks for making a start with it.

There's a lot going on in the second commit, so it would be great it could be broken up so that I can follow what's changing. Could you also add the "why" to the commit messages?

KludgeKML committed:
Replace unnecessary translation

- Remove translation of slugs in to_param
- Remove translation of division slug in view

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Ensure slug is not overwritten by data

- We always pass the slug into the division, and it's translated
  at that point if necessary, so stop it being overwritten by the
  slug in the data

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Remove obsolete method, force locale

- locale is usually only set for HTML, but for the
  json and ICS files here it should be set so that they
  can be returned in the requested language

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Allow calendar to pick up relevant datafile

- There's one datafile shared between Welsh and English paths,
  which in the original code is refered to as the scope (so all
  translations of the bank holidays data have the same scope,
  but different slugs).
- Allow the Calendar instance to retain the desired slug without
  having it translated away in the controller.
- Provide a class method that will translate the scope to the relevant
  scope, and a reader attribute of same on the instance.

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

KludgeKML committed:
Simplify scope

- Now that the Calendar model handles translating scope
  slug to a valid scope, we remove the scope method here
  and just refer to the parameter to make it easier to differentiate.

Co-authored-by: Ramya Vidapanakal <ramya.vidapanakal@digital.cabinet-office.gov.uk>

leenagupte reviewed (approved):
Thanks for breaking up the commits and adding the test. That made these changes much easier to follow.

That was a lot to unpick, so well done for figuring it all out. The use of "scope" is a bit confusing, but you've managed to make it easier to understand 🎉 

There's one really minor comment that's not a blocker, otherwise it all looks really good 👍

KludgeKML committed:
Remove Calendar Publishing task

- As of https://github.com/alphagov/special-route-publisher/pull/351,
  responsibility for publishing the calendars is now
  moved to special-route-publisher.

KludgeKML committed:
Remove CalendarContentItem

- without the publishing task, this is only used to support a
   test, so refactor the test and delete this class and its test
- FAQ Presenter only uses the title and description from the
  content item

KludgeKML commented on app/controllers/calendar_controller.rb:
```diff
@@ -31,7 +31,7 @@ def division
     respond_to do |format|
       format.json { render json: div }
-      format.ics { render plain: IcsRenderer.new(div.events, request.path).render }
+      format.ics { render plain: IcsRenderer.new(div.events, request.path, I18n.locale).render }
```
set_locale is what makes I18n.locale have the right values, so it's needed (and it does affect the json)

leenagupte reviewed (approved):
👍
